privacyidea_pam.py

```diff
@@ -114,7 +124,44 @@ def authenticate(self, password):
                     save_auth_item(self.sqlfile, self.user, serial, tokentype,
                                    auth_item)
                 else:
-                    rval = self.pamh.PAM_AUTH_ERR
+                    transaction_id = detail.get("transaction_id")
+                    challenge = None
+
+                    if "attributes" in detail:
+                        attributes = detail.get("attributes")
+                        if "u2fSignRequest" in attributes:
+                            syslog.syslog(syslog.LOG_DEBUG,
+                                    "Prompting for U2F authentication")
+# In case of U2F the $attributes looks like this:
+# {
+#     "img": "static/css/FIDO-U2F-Security-Key-444x444.png#012",
+#     "hideResponseInput" "1",
+#     "u2fSignRequest": {
+#         "challenge": "yji-PL1V0QELilDL3m6Lc-1yahpKZiU-z6ye5Zz2mp8",
+#         "version": "U2F_V2",
+#         "keyHandle": "fxDKTr6o8EEGWPyEyRVDvnoeA0c6v-dgvbN-6Mxc6XBmEItsw",
+#         "appId": "https://172.16.200.138"
+#     }
+# }
+                            challenge = """
+----- BEGIN U2F CHALLENGE -----
+%s
+%s
+%s
+----- END U2F CHALLENGE -----""" % (self.URL,
+                                    json.dumps(attributes["u2fSignRequest"]),
+                                    str(detail.get("message", "")))
+                    if transaction_id:
```

Comment:
Just from a logic standpoint I would first check for transaction_id, since transaction_id indicates, that we do a challenge_response.
Within the _transaction_id if-branch_, I would then check if it is U2F-Challenge and if not do "unsupported challenge". This way, we can add other challenges in the _transaction_id if-branch_ later, easily.